includes/class-wc-subscriptions-renewal-order.php

```diff
@@ -138,6 +138,11 @@ public static function maybe_record_subscription_payment( $order_id, $orders_old
 				if ( $is_failed_renewal_order ) {
 					do_action( 'woocommerce_subscriptions_paid_for_failed_renewal_order', wc_get_order( $order_id ), $subscription );
 				}
+			} elseif ( 'pending' === $orders_new_status && 'failed' === $orders_old_status && wcs_is_checkout_blocks_api_request() ) {
+				// The WooCommerce block checkout will update the order status from "failed" to "pending" right before attempting to process the payment.
+				// We need to flag this event so we can still trigger paid-for-failed-renewal-order actions once the order is paid.
+				$order->update_meta_data( '_was_failed_renewal_order', 'true' );
```

Comment:
IMO this kind of setting meta purely for operational/logical reasons isn't great, however I'm not sure of a better solution.

Once the order is set to pending, there's no way of knowing if the previous status was failed. You could read the order notes, however that is also fraught with issues and could be flakey.

If you have better idea, I'd be very happy to hear it. :)